### ISSUE TITLE
OnPush Change detection logic should not markAsCheckOnce when status is Detached

### ISSUE THREAD

**Issue by shlomiassaf** [angular/angular#9720]
**I'm submitting a ...**  (check one with "x")

```
[x] bug report
[x] feature request
```

**Current behavior**
When change detection strategy is set to OnPush you can not detach it.
Plunker: http://plnkr.co/edit/H4NZGy?p=preview

**Expected/desired behavior**
OnPush should respect detach().

**What is the motivation / use case for changing the behavior?**
Following @bennadel [blog post](http://www.bennadel.com/blog/3114-change-detection-strategy-appears-to-override-the-changedetectorref-in-angular-2-rc-3.htm) I was wondering why this happens.
I found that the `detectChangesInternal()` implementation in the view factory created by the code generator call `markAsCheckOnce()` on every change detection pass.

```
 if (changed) { self._appEl_11.componentView.markAsCheckOnce(); }
```

The code gen should only call `markAsCheckOnce()` when `cdMode !== ChangeDetectionStrategy.Detached`

A workaround would be to call `changeDetectionRef.detach()` inside `ngOnChanges` life cycle hook, this will revert the state back.
- **Angular version:** 2.0.0-rc.3

**Comment by shlomiassaf** [comment]
@vsavkin Is this a bug? feature?

Will it go into angular? if so roughly when?

Some labels are missing :)

**Comment by tbosch** [comment]
This is still a bug, see http://plnkr.co/edit/n4e7J4WGznOunEnew3ah?p=preview
